### ISSUE TITLE
Allow adding a callback to a list of futures

### ISSUE THREAD

**Issue by dhirschfeld** [dask/dask#1608]
As discussed on gitter I'm looking for a way to asynchronously handle the result of a parallel calculation represented by a list of futures e.g. to log either the success or failure of each task.

The current api allows you to chain calls and will automatically pass the _result_ of the tasks to the next function. If one of the tasks fails you get an exception in the calling process which prevents you handling the error asynchronously - you have to block in the calling process to handle the error

The `concurrent.futures` api has a [`Future.add_done_callback`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.add_done_callback) method which may be worthwhile implementing in `dask.distributed`

**Comment by dhirschfeld** [comment]
**Demo:**
```python
from numpy.random import rand
from distributed import Client, wait
import tornado


client = Client()

def func(x):
    from time import sleep
    if x < 0.1:
        raise Exception()
    sleep(x)
    return x

def post_process(futures):
    return [future.status for future in futures]
```
```
In [2]: futures = client.map(func, rand(20))
   ...: res = client.submit(post_process, futures)
   ...: res.result()
   ...: 
Traceback (most recent call last):

  File "<ipython-input-2-2652c88a48d0>", line 3, in <module>
    res.result()

  File "C:\Python\lib\site-packages\distributed\client.py", line 109, in result
    six.reraise(*result)

  File "C:\Python\lib\site-packages\six.py", line 685, in reraise
    raise value.with_traceback(tb)

  File "<ipython-input-1-ab12cbc30462>", line 11, in func
    raise Exception()

Exception
```
As can be seen above the `post_process` never gets called because of an exception in one of the tasks. Even if it did get called the `Client.submit` method would have passed the results of the futures rather than the futures themselves

**Comment by dhirschfeld** [comment]
If I try the suggested workaround:
```python
@tornado.gen.coroutine
def on_finished(future, callback):
    from distributed import wait
    yield wait(future)
    callback()
```
...I get a deadlock - the futures never get executed:
```python
def log_result(*futures):
    with open("test.log", 'w') as fstream:
        fstream.writelines('\n'.join('%r' % future for future in futures))
```
```python
In [18]: futures = client.map(func, rand(5))
    ...: client.loop.add_callback(on_finished, futures, log_result)

In [19]: futures
Out[19]: 
[<Future: status: pending, key: func-180c683d583878179ddfd4d5dc5cdb90>,
 <Future: status: pending, key: func-7376f93f6c5292e37c2117565acd18e6>,
 <Future: status: pending, key: func-e30ad896473972868ddd4213c516eaf9>,
 <Future: status: pending, key: func-35aed41d26bd8af134b3da48f3709f69>,
 <Future: status: pending, key: func-a639b6ba8a4db971f311946292ccd76d>]

In [20]: client.gather(futures)  # WAIT FOREVER!!!
```

**Comment by dhirschfeld** [comment]
If I instead wait for the futures to complete before adding the callback the callback appears to not be called:
```
In [3]: futures = client.map(func, rand(5))
   ...: res = wait(futures)

In [4]: futures
Out[4]: 
[<Future: status: finished, type: float64, key: func-9dcb25bcda1613ded1a588cacbcef778>,
 <Future: status: finished, type: float64, key: func-2ad77f2c4d5aad87f88305aaefa66b35>,
 <Future: status: finished, type: float64, key: func-8d6bc30fa11e6d70a53182116a856021>,
 <Future: status: finished, type: float64, key: func-c5873711dd45bba262b771b37e33f6b9>,
 <Future: status: finished, type: float64, key: func-d66632c57baa4722a3cfe84195a4f951>]

In [5]: client.loop.add_callback(on_finished, futures, log_result)

In [6]: !cat test.log
cat: test.log: No such file or directory  <--- CALLBACK NEVER GOT EXECUTED!
```
...but manually calling the `log_results` function works as expected:
```python

In [10]: log_result(futures)

In [11]: !cat test.log
<Future: status: finished, type: float64, key: func-9dcb25bcda1613ded1a588cacbcef778>
<Future: status: finished, type: float64, key: func-2ad77f2c4d5aad87f88305aaefa66b35>
<Future: status: finished, type: float64, key: func-8d6bc30fa11e6d70a53182116a856021>
<Future: status: finished, type: float64, key: func-c5873711dd45bba262b771b37e33f6b9>
<Future: status: finished, type: float64, key: func-d66632c57baa4722a3cfe84195a4f951>
```

**Comment by mrocklin** [comment]
You need to use the underscore version, `_wait`, which is a tornado coroutine rather than a blocking function.  This requires understanding of Tornado coroutines and asynchronous programming.  Here is a good link:

http://www.tornadoweb.org/en/stable/coroutine.html


### Example

```python
In [1]: from distributed import Client

In [2]: c = Client()

In [3]: from distributed.client import _wait

In [4]: from tornado import gen

In [5]: def div(x, y):
   ...:     return x / y
   ...:

In [6]: future = c.submit(div, 1, 0)

In [7]: @gen.coroutine
   ...: def print_status(future):
   ...:     yield _wait(future)
   ...:     print(future.status)
   ...:

In [8]: c.loop.add_callback(print_status, future)
 error
```

**Comment by dhirschfeld** [comment]
Sorry, my tornado-fu is fairly non-existent :(

Thanks's a lot, I can confirm the above works as expected! :)

 I did checkout the docs and managed to get it working with `add_future` but will go with your solution above which seems less hacky
```python
@tornado.gen.coroutine
def on_finished(futures):
    yield wait(futures)

client.loop.add_future(on_finished(futures), log_result)
```

**Comment by mrocklin** [comment]
See solution in https://github.com/dask/distributed/pull/560

**Comment by dhirschfeld** [comment]
I've managed to deadlock it again :/

The below callback works fine *if* I have `exc future.exception()` commented out:
```python
def func(x):
    from time import sleep
    sleep(x)
    1/0


@tornado.gen.coroutine
def on_finished(futures, callback):
    yield _wait(futures)
    callback(futures)


def callback(futures):
    for future in futures:
        #exc = future.exception()
        print(future)


futures = client.map(func, rand(5))
client.loop.add_callback(on_finished, futures, callback)
print("END!")
```
![distributed](https://cloud.githubusercontent.com/assets/881019/19065513/55bf558a-8a57-11e6-9133-73443e2f896d.gif)

If however I try to access the `future.exception()` method the callback hangs. I don't suppose you know what I'm doing wrong this time?

**Comment by dhirschfeld** [comment]
```python
def callback(futures):
    print('Entered callback', flush=True)
    for future in futures:
        exc = future.exception()
        print(future, flush=True)
    print('Left callback', flush=True)
```
![distributed2](https://cloud.githubusercontent.com/assets/881019/19065705/5680b788-8a58-11e6-94e2-b2b654831016.gif)

**Comment by mrocklin** [comment]
You can't use any of the blocking API inside a coroutine.  You'll need to use the `_exception` method rather than `exception`.

**Comment by mrocklin** [comment]
You could also use standard concurrent.futures and threads and use the blocking API as normal

```python
from concurrent.futures import ThreadPoolExecutor
e = ThreadPoolExecutor()

def wait_and_run(futures, callback):
    wait(futures)
    callback(futures)

e.submit(wait_and_run, futures, callback)
```

This uses more standard technology and will probably involve fewer speedbumps around Tornado.

**Comment by dhirschfeld** [comment]
Thanks for your patience @mrocklin - using the `_exception` method seems to work great. 

Using `concurrent.futures` as well as `distributed` in the same workflow seems a little hacky, and would also tie me to the local machine IIUC.

As a little background I'm trying to implement some ETL workflows that I've previously done in Celery so I'm probably pushing distributed into territory it's not specifically designed for (as evidenced by the use of *"private"* methods!) In these workflows it's important to be able to add custom error handling so the default of chaining futures by passing `future.result()` doesn't work well for this. 

Whilst I'm happy that I can do what I'm after with `on_finished`/`client.loop.add_callback` it might be nice to provide a default way of adding `on_error`/`on_succcess` handlers to a call to `client.submit` and `client.map` - i.e. make `on_error`/`on_success` new keyword arguments to `submit` and `map` and have them append a callback to the futures.

In the example below, `on_finished` would become an internal implementation detail in `distributed` and instead of a separate function `parallel_map` the `client.map` method would implement the `on_success`/`on_error` handling directly:
```python
@tornado.gen.coroutine
def on_finished(futures, on_success=None, on_error=None):
    yield _wait(futures)
    ERROR = any(
        future.status == 'error'
        for future in futures
    )
    if ERROR:
        if on_error is not None:
            on_error(futures)
    else:
        if on_success is not None:
            on_success(futures)


def parallel_map(client, func, *iterables, **kwargs):
    on_success = kwargs.pop('on_success', None)
    on_error = kwargs.pop('on_error', None)
    futures = client.map(func, *iterables, **kwargs)
    if on_success is not None or on_error is not None:
        client.loop.add_callback(on_finished, futures, on_success, on_error)
    return futures
```
```python
In [59]: futures = parallel_map(client, lambda x: 1/0, randn(5), on_success=lambda futs: print('SUCCESS!'), on_error=lambda futs: print('ERROR!'))
ERROR!

In [60]: futures = parallel_map(client, lambda x: x, randn(5), on_success=lambda futs: print('SUCCESS!'), on_error=lambda futs: print('ERROR!'))
SUCCESS!
```

**Comment by dhirschfeld** [comment]
Similarly for `submit`:
```python
@tornado.gen.coroutine
def on_finished(futures, on_success=None, on_error=None):
    yield _wait(futures)
    try:
        it = iter(futures)
    except TypeError:
        it = iter([futures])
    ERROR = any(
        future.status == 'error'
        for future in it
    )
    if ERROR:
        if on_error is not None:
            on_error(futures)
    else:
        if on_success is not None:
            on_success(futures)


def parallel_submit(client, func, *args, **kwargs):
    on_success = kwargs.pop('on_success', None)
    on_error = kwargs.pop('on_error', None)
    future = client.submit(func, *args, **kwargs)
    client.loop.add_callback(on_finished, future, on_success, on_error)
    return future
```
```python
In [68]: parallel_submit(client, lambda: 1, on_success=lambda fut: print('SUCCESS!'), on_error=lambda fut: print('ERROR!'))
Out[68]: <Future: status: pending, key: <lambda>-b88df71dc0bda4e0520419ab30edb582>
SUCCESS!

In [69]: parallel_submit(client, lambda: 1/0, on_success=lambda fut: print('SUCCESS!'), on_error=lambda fut: print('ERROR!'))
Out[69]: <Future: status: pending, key: <lambda>-ea2947d44eeb6de94e714d0ab43f9098>
ERROR!

In [70]: _68
Out[70]: <Future: status: finished, type: int, key: <lambda>-b88df71dc0bda4e0520419ab30edb582>

In [71]: _69
Out[71]: <Future: status: error, key: <lambda>-ea2947d44eeb6de94e714d0ab43f9098>
```

**Comment by dhirschfeld** [comment]
Alternate implementation using `add_done_callback` from dask/distributed#560:
```python
def parallel_map(client, func, *iterables, **kwargs):
    on_success = kwargs.pop('on_success', None)
    on_error = kwargs.pop('on_error', None)
    futures = client.map(func, *iterables, **kwargs)
    if on_success is not None or on_error is not None:
        future = client.submit(lambda arg: arg, futures)

        def on_finished(future):
            if future.status == 'error':
                if on_error is not None:
                    on_error(futures)
            else:
                if on_success is not None:
                    on_success(futures)

        future.add_done_callback(on_finished)

    return future
```

**Comment by dhirschfeld** [comment]
Actually the above won't work as the `on_finished` callback will be called on the *first* exception when not all futures may have completed so you would still need to wait on the futures before being able to process them all

**Comment by mrocklin** [comment]
cc @minrk any thoughts on how to cleanly extend the Futures API to support `on_exception` etc?

**Comment by minrk** [comment]
The [asyncio api](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait) defines constants to be used as flags in collective operations to determine when callbacks fire, if you are interested in controlling how/when callbacks run:

> - FIRST_COMPLETED	The function will return when any future finishes or is cancelled.
> - FIRST_EXCEPTION	The function will return when any future finishes by raising an exception. If no future raises an exception then it is equivalent to ALL_COMPLETED.
> - ALL_COMPLETED The function will return when all futures finish or are cancelled.

You could follow a similar pattern for on_error, on_finished:

```python
futures = client.map(func, *iterables)
# wait for everything OR first exception if one happens
yield client.wait(futures, return_when=FIRST_EXCEPTION)
# wait for everything (all exceptions, if many)
yield client.wait(futures, return_when=ALL_COMPLETED)
# wait for the first result, success or failure
done,  pending = yield client.wait(futures, return_when=FIRST_COMPLETED)
first_result = done[0]
```

`on_error` as a kwarg to map could be able to be implemented in the same way, as long as you have `client.wait` with those semantics, possibly including an additional flag for whether `on_error` should be called on the first error, or waiting for them all, e.g.

```python
def print_errors(futures):
    failed = [ f for f in futures if f.exception() ]
    for f in failed:
        print(f.exception())

# default: on_error_when=FIRST_EXCEPTION
client.map(func, seq, on_error=print_errors, on_error_when=ALL_COMPLETED) 
```

**Comment by jcrist** [comment]
This appears to have been fixed in https://github.com/dask/distributed/pull/560. Good to close?

**Comment by dhirschfeld** [comment]
IIUC `#560` added `add_done_callback` to the `Future` object to mimic the `concurrent.futures` interface. There's still no simple (one-line) way to do the same for a list of futures - it's possible, but not obvious.

I've had some thoughts about additions to the api to better support ETL workflows but haven't had the chance to put them down here. I'm happy to open a new issue for that though if you'd like to close this one...

**Comment by jcrist** [comment]
Apologies then, I didn't take the time to fully read through the discussion here. I'm in no rush to close it, just wanted to make sure it hadn't gone stale. Since this has to do with distributed instead of dask, would you mind moving what remains for this issue to [an issue in that repo](https://github.com/dask/distributed/issues) (and then closing this one)?

**Comment by dhirschfeld** [comment]
Discussion continued in dask/distributed#612

**Comment by jakirkham** [comment]
Did you find a satisfactory solution to this problem @dhirschfeld?

**Comment by dhirschfeld** [comment]
I found a solution which works but is pretty ugly. I'm revisiting this problem again shortly so will open new issues if there are still some rough edges in the api.

My thoughts are to try and use the new async client and as such I'm interested in the discussion around #1663.
